src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceService.ChangeNamespaceCodeAction.cs

```diff
@@ -0,0 +1,526 @@
+                }
+
+                foreach(var container in containers)
+                {
+                    editor.TrackNode(container);
+                }
+
+                document = editor.GetChangedDocument();
+                root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                return (document, containers.Select(c => root.GetCurrentNode(c)).ToImmutableArray());
+            }
+
+            private async Task<Solution> RemoveUnnecessaryImportsAsync(
+                Solution solution, 
+                ImmutableArray<DocumentId> ids, 
+                ImmutableArray<string> names, 
+                CancellationToken cancellationToken)
+            {
+                var LinkedDocumentsToSkip = PooledHashSet<DocumentId>.GetInstance();
+                foreach (var id in ids)
+                {
+                    if (LinkedDocumentsToSkip.Contains(id))
+                    {
+                        continue;
+                    }
+
+                    var document = solution.GetDocument(id);
+                    LinkedDocumentsToSkip.AddRange(document.GetLinkedDocumentIds());
+
+                    document = await RemoveUnnecessaryImportsAsync(
+                        document, 
+                        CreateImports(document, names, withFormatterAnnotation: false), 
+                        cancellationToken)
+                        .ConfigureAwait(false);
+                    solution = document.Project.Solution;
+                }
+
+                LinkedDocumentsToSkip.Free();
+                return solution;
+            }
+
+            private async Task<Document> RemoveUnnecessaryImportsAsync(
+                Document document, 
+                IEnumerable<SyntaxNode> importsToRemove,
+                CancellationToken cancellationToken)
+            {
+                var removeImportService = document.GetLanguageService<IRemoveUnnecessaryImportsService>();
+                return await removeImportService.RemoveUnnecessaryImportsAsync(
+                    document,
+                    import => importsToRemove.Any(importToRemove => importToRemove.IsEquivalentTo(import, topLevel: false)),
+                    cancellationToken)
+                    .ConfigureAwait(false);
+            }
+
+            private async Task<Document> AddImportsInContainersAsync(
+                Document document,
+                IAddImportsService addImportService,
+                ImmutableArray<SyntaxNode> containers, 
+                ImmutableArray<string> names,
+                bool placeSystemNamespaceFirst, 
+                CancellationToken cancellationToken)
+            {
+                // Sort containers based on their span start, to make the result of 
+                // adding imports deterministic. This is mostly to make unit tests pass 
+                // and might not necessary for the correctness of the refactoring. 
```

Comment:
note: we shoudl always try to be deterministic in what we do, not jsut for tests.  It can otherwise lead to very confusing behavior for users.

Reply:
Fixed comments

---
In reply to: [230476795](https://github.com/dotnet/roslyn/pull/30156#discussion_r230476795) [](ancestors = 230476795)